typing/typecore.ml

```diff
@@ -5582,16 +5619,36 @@ let report_error ~loc env = function
            fprintf ppf "This expression has type")
         (function ppf ->
            fprintf ppf "but an expression was expected of type");
-  | Apply_non_function typ ->
-      begin match get_desc typ with
+  | Apply_non_function { func_name; funct; extra_arg_loc; previous_arg_loc } ->
+      begin match get_desc funct with
         Tarrow _ ->
-          Location.errorf ~loc
-            "@[<v>@[<2>This function has type@ %a@]\
-             @ @[It is applied to too many arguments;@ %s@]@]"
-            Printtyp.type_expr typ "maybe you forgot a `;'.";
+          let app_loc =
+            (* Span the application, including the extra argument. *)
+            { Location.loc_start = loc.Location.loc_start;
+              loc_end = extra_arg_loc.loc_end;
+              loc_ghost = false }
+          and tail_loc =
+            (* Possible location for a ';'. *)
+            let loc_start = previous_arg_loc.loc_end in
+            { Location.loc_start;
+              loc_end = { loc_start with pos_cnum = loc_start.pos_cnum + 1 };
+              loc_ghost = false }
+          in
+          let sub = [
+            Location.msg ~loc:tail_loc "Hint: Did you forget a ';'?";
+            Location.msg ~loc:extra_arg_loc "This argument is not expected.";
+          ]
+          and pp_func_name ppf = function
+            | Some name -> Format.fprintf ppf "The function '%s'" name
+            | None -> Format.fprintf ppf "This function"
+          in
+          Location.errorf ~loc:app_loc ~sub
+            "@[<v>@[<2>%a has type@ %a@]\
+             @ It is applied to too many arguments@]"
+            pp_func_name func_name Printtyp.type_expr funct;
```

Comment:
It wouldn't be the first error message written that way, so it's entirely out. It's not easy to split further because most of it is computing the locations we want to show specifically for this case, so I kept it as one block.